src/network/network_crypto.h

```diff
@@ -0,0 +1,183 @@
+ */
+class NetworkEncryptionHandler {
+public:
+	virtual ~NetworkEncryptionHandler() {}
+
+	/**
+	 * Get the size of the MAC (Message Authentication Code) used by the underlying encryption protocol.
+	 * @return The size, in bytes, of the MACs.
+	 */
+	virtual size_t MACSize() const = 0;
+
+	/**
+	 * Decrypt the given message in-place, validating against the given MAC.
+	 * @param mac The message authentication code (MAC).
+	 * @param message The location of the message to decrypt.
+	 * @return Whether decryption and authentication/validation of the message succeeded.
+	 */
+	virtual bool Decrypt(std::span<std::uint8_t> mac, std::span<std::uint8_t> message) = 0;
+
+	/**
+	 * Encrypt the given message in-place, and write the associated MAC.
+	 * @param mac The location to write the message authentication code (MAC) to.
+	 * @param message The location of the message to encrypt.
+	 */
+	virtual void Encrypt(std::span<std::uint8_t> mac, std::span<std::uint8_t> message) = 0;
+};
+
+/** The authentication method that can be used. */
+enum NetworkAuthenticationMethod : uint8_t {
+	NETWORK_AUTH_METHOD_X25519_PLAIN, ///< No actual authentication is taking place, just perform a x25519 key exchange.
+	NETWORK_AUTH_METHOD_X25519_PAKE, ///< Authentication using x25519 password-authenticated key agreement.
+	NETWORK_AUTH_METHOD_END, ///< Must ALWAYS be on the end of this list!! (period)
+};
+
+/** The mask of authentication methods that can be used. */
+using NetworkAuthenticationMethodMask = uint16_t;
+
+/**
+ * Base class for cryptographic authentication handlers.
+ */
+class NetworkAuthenticationHandler {
+public:
+	virtual ~NetworkAuthenticationHandler() {}
+
+	/**
+	 * Get the name of the handler for debug messages.
+	 * @return The name of the handler.
+	 */
+	virtual std::string_view GetName() = 0;
+
+	/**
+	 * Get the method this handler is providing functionality for.
+	 * @return The \c NetworkAuthenticationMethod.
+	 */
+	virtual NetworkAuthenticationMethod GetAuthenticationMethod() = 0;
+
+	/**
+	 * Create a \a NetworkEncryptionHandler to encrypt or decrypt messages from the client to the server.
+	 * @return The handler for the client to server encryption.
+	 */
+	virtual std::unique_ptr<NetworkEncryptionHandler> CreateClientToServerEncryptionHandler() const = 0;
+
+	/**
+	 * Create a \a NetworkEncryptionHandler to encrypt or decrypt messages from the server to the client.
+	 * @return The handler for the server to client encryption.
+	 */
+	virtual std::unique_ptr<NetworkEncryptionHandler> CreateServerToClientEncryptionHandler() const = 0;
+};
+
+/**
+ * Base class for client side cryptographic authentication handlers.
+ */
+class NetworkAuthenticationClientHandler : public NetworkAuthenticationHandler {
+public:
+	/** The processing result of receiving a request. */
+	enum RequestResult {
+		AWAIT_USER_INPUT, ///< We have requested some user input, but must wait on that.
```

Comment:
This name feels very generic. But I guess it is really specific: `AWAIT_USER_PASSWORD_INPUT` or something?

Reply:
Currently it is, indeed, really specific for awaiting a password. But if at some point someone wants to introduce 2FA, where you also have to some on-time-token, then it wouldn't be a password any more but the further handling by the network code would be the same.

With `AWAIT_USER_INPUT` the network handling code at the client is essentially done, and at some point there will be a callback that triggers sending the response. For this the network handling code already passed a callback structure when preparing the authentication, so that arranges everything that's needed.

With `IMMEDIATE_RESPONSE` there is no need to ask the user for something, and it can immediately trigger sending that response.